dsa/src/signing_key.rs

```diff
@@ -63,31 +68,34 @@ impl SigningKey {
     ///
     /// If you decide to clone this value, please consider using [`Zeroize::zeroize`](::zeroize::Zeroize::zeroize()) to zero out the memory after you're done using the clone
     #[must_use]
-    pub fn x(&self) -> &BigUint {
+    pub fn x(&self) -> &NonZero<BoxedUint> {
         &self.x
     }
 
     /// Sign some pre-hashed data
     fn sign_prehashed(
         &self,
-        (k, inv_k): (BigUint, BigUint),
+        (k, inv_k): (BoxedUint, BoxedUint),
         hash: &[u8],
     ) -> signature::Result<Signature> {
         let components = self.verifying_key().components();
         let (p, q, g) = (components.p(), components.q(), components.g());
         let x = self.x();
 
-        let r = g.modpow(&k, p) % q;
+        let params = BoxedMontyParams::new(Odd::new(k).unwrap());
+        let form = BoxedMontyForm::new((**g).clone(), params);
+        let r = NonZero::new(form.pow(p).retrieve() % q).unwrap();
```

Comment:
Looks like the test failure is occurring on line #85, ostensibly due to the `unwrap` and an even `k`.

I'm not sure why `k` would be even. Perhaps there's an issue in how you're implementing RFC6979? But either way, it would be good to return `signature::Error` in that case rather than panicking.

You might also be able to leverage the `rfc6979` crate's implementation of the algorithm now as well, rather than just it's implementation of HMAC-DRBG: https://github.com/RustCrypto/signatures/blob/ef35987cd1bca5029a4d1d2c6bf3e31ddea0da38/rfc6979/src/lib.rs#L77-L113

Reply:
Yep. I'm also asking myself how it could even be even. And yes, the `.unwrap()`s are part of the "ugly code" I mentioned at the top of the PR, heh.  
I am intending to clean it all up with proper error handling and everything after getting it to work.

Reply:
It definitely is an issue with my own implementation but only ever since I moved to `crypto-bigint` since it's odd in the `num-bigint-dig` version of the code.  
I'll attempt to move it over to the newly exposed API.

Reply:
Now to adjusting the capacity of the generator `BoxedUint`..